File: src/main/java/com/cjw/reactivecommunityproject/common/security/service/impl/JwtServiceImpl.java
package com.cjw.reactivecommunityproject.common.security.service.impl;

import com.cjw.reactivecommunityproject.common.exception.model.RcBaseException;
import com.cjw.reactivecommunityproject.common.exception.model.RcCommonErrorMessage;
import com.cjw.reactivecommunityproject.common.security.exception.SecurityErrorMessage;
import com.cjw.reactivecommunityproject.common.security.exception.SecurityException;
import com.cjw.reactivecommunityproject.common.security.model.SecurityAccessJwt;
import com.cjw.reactivecommunityproject.common.security.model.SecurityJwtPayload;
import com.cjw.reactivecommunityproject.common.security.service.JwtService;
import com.cjw.reactivecommunityproject.server.cache.custom.service.CacheCustomService;
import com.cjw.reactivecommunityproject.web.auth.exception.AuthException;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.jsonwebtoken.ExpiredJwtException;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.MalformedJwtException;
import io.jsonwebtoken.UnsupportedJwtException;
import io.jsonwebtoken.security.Keys;
import io.jsonwebtoken.security.SignatureException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.apache.commons.lang3.math.NumberUtils;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.stereotype.Service;

import javax.crypto.SecretKey;
import java.util.Base64;
import java.util.Date;

@Slf4j
@Service
@RequiredArgsConstructor
public class JwtServiceImpl implements JwtService {
    private final ObjectMapper objectMapper;
    private final RedisTemplate<String, Object> redisTemplate;
    private final CacheCustomService cacheCustomService;


    private Long getTokenExpiresByCommonEnvCode(String tokenType) {
        var envcode = cacheCustomService.getCustomCommonEnvCode(StringUtils.join("rc.jwt.", tokenType));
        if (envcode == null) {
            throw new AuthException(RcCommonErrorMessage.NOT_FOUND_ENV_CODE);
        }
        if (!NumberUtils.isDigits(envcode.getValue())) {
            throw new AuthException(RcCommonErrorMessage.INVALID_ENV_CODE);
        }

        return NumberUtils.toLong(envcode.getValue());
    }

    private Long getAccessTokenExpiresByCommonEnvCode() {
        return getTokenExpiresByCommonEnvCode("access.token.expires.minutes");
    }

    private Long getRefreshTokenExpiresByCommonEnvCode() {
        return getTokenExpiresByCommonEnvCode("refresh.token.expires.minutes");
    }

    private String getSecretKeyByCommonEnvCode() {
        var envcode = cacheCustomService.getCustomCommonEnvCode("rc.jwt.secret.key");
        if (envcode == null) {
            throw new AuthException(RcCommonErrorMessage.NOT_FOUND_ENV_CODE);
        }
        if (StringUtils.isBlank(envcode.getValue())) {
            throw new AuthException(RcCommonErrorMessage.INVALID_ENV_CODE);
        }

        return envcode.getValue();
    }

    private String getUserUidByJwtPayload(String token) {
        String[] parts = StringUtils.split(token, ".");
        if (parts.length != 3) {
            throw new SecurityException(SecurityErrorMessage.INVALID_TOKEN_STRUCT);
        }

        String payload = parts[1];
        byte[] decodedPayload = Base64.getUrlDecoder().decode(payload);
        try {
            var jwtPayloadVO = objectMapper.readValue(new String(decodedPayload), SecurityJwtPayload.class);
            return jwtPayloadVO.sub();
        } catch (JsonProcessingException ex) {
            throw new SecurityException(SecurityErrorMessage.INVALID_TOKEN_PAYLOAD);
        }
    }

    private SecretKey getSecretKey(String userUid) {
        String salt;
        if (StringUtils.isBlank(userUid)) {
            salt = "";
        } else {
            salt = String.valueOf(redisTemplate.opsForValue().get(userUid + ".salt"));
        }
        if (StringUtils.equalsIgnoreCase(salt, "null")) {
            throw new SecurityException(SecurityErrorMessage.NOT_FOUND_USER_SALT);
        }

        var secretKey = this.getSecretKeyByCommonEnvCode() + "." + salt;
        return Keys.hmacShaKeyFor(secretKey.getBytes());
    }

    private String createToken(SecurityAccessJwt securityAccessJwt, Long expiresMinutes) {
        return Jwts.builder()
                .subject(securityAccessJwt.userUid())
                .claim("role", securityAccessJwt.roleUid())

                .issuedAt(new Date())
                .expiration(new Date(new Date().getTime() + expiresMinutes))

                .signWith(this.getSecretKey(securityAccessJwt.userUid()))
                .compact();
    }


    @Override
    public String createAccessToken(SecurityAccessJwt securityAccessJwt) {
        return createToken(securityAccessJwt, this.getAccessTokenExpiresByCommonEnvCode() * 1000 * 60);
    }

    @Override
    public String createRefreshToken(String userUid) {
        return createToken(SecurityAccessJwt.builder()
                .userUid(userUid)
                .roleUid(null)
                .build(), this.getRefreshTokenExpiresByCommonEnvCode() * 1000 * 60);
    }

    @Override
    public Boolean validateToken(String token) {
        return this.getClaims(token) != null;
    }

    @Override
    public SecurityAccessJwt getClaims(String token) {
        try {
            var userUid = this.getUserUidByJwtPayload(token);
            var claims = Jwts.parser()
                    .verifyWith(this.getSecretKey(userUid))
                    .build()
                    .parseSignedClaims(token);
            return SecurityAccessJwt.builder()
                    .userUid(
                            claims.getPayload().getSubject()
                    )
                    .roleUid(
                            claims.getPayload().get("role", Integer.class)
                    )
                    .build(); // 토큰이 유효함
        } catch (SignatureException e) {
            log.warn("JwtServiceImpl.getClaims(): Invalid Jwt Signature");
        } catch (MalformedJwtException e) {
            log.warn("JwtServiceImpl.getClaims(): Invalid Jwt Token");
        } catch (ExpiredJwtException e) {
            log.warn("JwtServiceImpl.getClaims(): Expires Jwt Token");
        } catch (UnsupportedJwtException e) {
            log.warn("JwtServiceImpl.getClaims(): Unsupported Jwt Token");
        } catch (IllegalArgumentException e) {
            log.warn("JwtServiceImpl.getClaims(): Jwt Token is empty");
        } catch (RcBaseException e) {
            log.warn("JwtServiceImpl.getClaims(): {}", e.getErrorMessage());
        } catch (Exception e) {
            log.error("JwtServiceImpl.getClaims(): {}", e.getMessage(), e);
        }
        return null;
    }
}

File: src/main/kotlin/com/devcjw/reactivecommunity/auth/service/impl/JwtServiceImpl.kt
package com.devcjw.reactivecommunity.auth.service.impl

import com.devcjw.reactivecommunity.auth.model.RcUserJwtClaims
import com.devcjw.reactivecommunity.auth.service.JwtService
import io.github.oshai.kotlinlogging.KotlinLogging
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import jakarta.annotation.PostConstruct
import lombok.RequiredArgsConstructor
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.util.*
import javax.crypto.SecretKey
import javax.crypto.spec.SecretKeySpec

@Service
@RequiredArgsConstructor
class JwtServiceImpl : JwtService {
    private val logger = KotlinLogging.logger {}

    private lateinit var jwtKey: SecretKey

    @Value("\${rc.jwt.access-token-expires-minute}")
    private val accessTokenExpiresMinute: Long? = null

    @Value("\${rc.jwt.refresh-token-expires-day}")
    private val refreshTokenExpiresDay: Long? = null

    @Value("\${rc.jwt.secret-key}")
    private val secretKey: String? = null

    @PostConstruct
    fun init() {
        jwtKey = SecretKeySpec(secretKey?.toByteArray(), SignatureAlgorithm.HS256.jcaName)
    }

    private fun getClaimsFormToken(token: String) = Jwts.parserBuilder()
        .setSigningKey(jwtKey)
        .build()
        .parseClaimsJws(token)

    private fun createToken(uid: String, level: Long, expiresTime: Long): String {
        val claims = Jwts.claims().apply {
            this["uid"] = uid
            this["level"] = level
        }

        val now = Date()
        val expires = Date(now.time + expiresTime)

        return Jwts.builder()
            .setClaims(claims)
            .setIssuedAt(now)
            .setExpiration(expires)
            .signWith(jwtKey)
            .compact()
    }

    override fun createAccessToken(uid: String, level: Long): String {
        logger.info { "create access token : uid => $uid, level => $level" }
        return createToken(uid, level, (accessTokenExpiresMinute?.times(1000) ?: 0) * 60)
    }

    override fun createRefreshToken(uid: String, level: Long): String {
        logger.info { "create refresh token : uid => $uid, level => $level" }
        return createToken(uid, level, (refreshTokenExpiresDay?.times(1000) ?: 0) * 60 * 60 * 24)
    }

    override fun validateToken(token: String): Boolean {
        return try {
            val claims = getClaimsFormToken(token)
            !claims.body.expiration.before(Date())
        } catch (ex: Exception) {
            logger.error(ex) { "validate Token error" }
            false
        }
    }

    override fun getRcUser(token: String): RcUserJwtClaims {
        val claims = getClaimsFormToken(token)
        return RcUserJwtClaims(claims.body["uid"].toString(), claims.body["level"].toString().toLong())
    }


}